Sources/App/Views/ViewModelQueries/PackageShowViewQuery.swift

```diff
@@ -1,32 +1,77 @@
+            .init(name: "Kevin Harwood", url: "https://github.com/kcharwood"),
+            .init(name: "186 other contributors", url: "https://github.com/Alamofire/Alamofire/graphs/contributors"),
         ],
           history: .init(
             since: "over 5 years",
             commits: .init(name: "1,433 commits",
-                           url: URL(string: "https://github.com/Alamofire/Alamofire/commits/master")!),
+                           url: "https://github.com/Alamofire/Alamofire/commits/master"),
             releases: .init(name: "79 releases",
-                            url: URL(string: "https://github.com/Alamofire/Alamofire/releases")!)
+                            url: "https://github.com/Alamofire/Alamofire/releases")
         ),
           activity: .init(
             openIssues: .init(name: "27 open issues",
-                              url: URL(string: "https://github.com/Alamofire/Alamofire/issues")!),
+                              url: "https://github.com/Alamofire/Alamofire/issues"),
             pullRequests: .init(name: "5 open pull requests",
-                                url: URL(string: "https://github.com/Alamofire/Alamofire/pulls")!),
+                                url: "https://github.com/Alamofire/Alamofire/pulls"),
             lastPullRequestClosedMerged: "6 days ago"),
           products: (libraries: 3, executables: 1)
     )
 }
+
+
+extension PackageShowView.Model {
+    static func query(database: Database, packageId: Package.Id) -> EventLoopFuture<Self> {
+        Package.query(on: database)
+            .with(\.$repositories)
+            .with(\.$versions) { $0.with(\.$products) }
+            .filter(\.$id == packageId)
+            .first()
+            .unwrap(or: Abort(.notFound))
+            .map { p in
+                // TODO: consider which missing values take defaults and which ones error out
+                // Or do we make view model fields optional and let display deal with it?
+                // If for instance we have just a package entry with no repository, we should
+                // probably error out (because too many fields would be blank).
+                Self.init(title: p.name ?? "–",
+                          url: p.url,
+                          license: p.repository?.license ?? .other,
+                          summary: p.repository?.summary ?? "",
+                          authors: [],
+                          history: .init(since: "foo",
+                                         commits: .init(name: "commits", url: "1"),
+                                         releases: .init(name: "releases", url: "2")),
+                          activity: .init(
+                            openIssues: .init(name: "issues", url: "3"),
+                            pullRequests: .init(name: "PRs", url: "4"),
+                            lastPullRequestClosedMerged: "XX days ago"),
+                          products: (libraries: 17, executables: 42))
+        }
```

Comment:
Dave, I'm thinking of making some of the model fields optional to allow us to chip away at the query while not everything is there yet and to better handle error cases where data is partially missing. (See also the TODO comment just above this initialiser).

So for instance if we have

```
struct Model {
  ...
  let history: History?
  ...
}
```

I'd leave this `nil` in the query and then in the display part we could conditionally leave out the whole section if data is lacking.

Does that make sense?

The alternative would be to leave the fields out until we can populate them but then we'd have to comment out the display part.

Reply:
Yes, it's better for the views to be able to deal with optional data. That way, if we can't get data for a specific package for any reason the page can still be displayed, just without that element.

Reply:
Excellent. Later on we can look at maybe returning `nil` from the model query if too many things are missing but that's not something we need to deal with right now.

Reply:
Ok, that's updated and good to go!